mkldnn:Conv weight optimization

sreekanth-yalachigere opened the pull request:
Filter data is constant for all iterations and currently we are reordering filter memory layout in every iteration. By saving filter memory layout in execution provider, we can avoid 53 data reorders per iterations on Resnet50 and reduce the latency.

Implemented thread safe weights update.

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -401,23 +401,35 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
+        filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
+        {
+          // make assignment threadsafe
+          std::lock_guard<std::mutex> lock(mutex_);
```
@jywu-msft will move lock to SetWeightMemory() of provider.

Re Atomic: are you suggesting to have one thread do the atomic operation of "checking + Reorder + Setting of all weights"?

jywu-msft replied:
we only want 1 thread to do the weights reordering.
with your existing code, multiple threads could do the reordering since you are only synchronizing the setting of the map.
alternatively, maybe you can try using std::call_once() instead of using a mutex.

sreekanth-yalachigere replied:
@jywu-msft can you please take a look?

created new function Reorder_weights in Conv (too many parameters) so that I can call std::call_once but couldn't figure out how to pass parameters to call_once args. 

Created Get_weights_memory and Set_weights_memory and moved lock and mutex to provider (for now, Will remove after call_once implementation)

jywu-msft replied:
did you test these changes? i saw failures in Conv unit tests when building these changes.
re: std::call_once, it would be something like

std::call_once(flag1, &Conv::Reorder_weights, this, context, src_dims_mkl, filter_dims_mkl, bias_dims_mkl, dst_dims_mkl, strides_mkl, dilations_mkl, padding_left_mkl, padding_right_mkl, filter_format, filter_dims_mkl);

sreekanth-yalachigere replied:
@jywu-msft 

`std::call_once` didn't work and it's called once in the process. Only first test passed and rest of the tests failed. `perf_test `(Resnet50) also fails.

I've moved the call_once implementation to new [branch](https://github.com/sreekanth-yalachigere/onnxruntime/blob/f2bf757d972d0ded02fbdd5f57bd5fe0762e3c66/onnxruntime/core/providers/mkldnn/nn/conv.cc#L435), . 

Here, I am using mutex for thread-safe implementation. 

```
check memory map
if (empty)
lock(){
  check memory map // did last thread re-order and save?
  if (empty){
     re-order
     save in weight-name and memory map
  }
}
```

jywu-msft replied:
lock should be owned by the provider, as the map is also owned by provider. 
as currently implemented, you can still have multiple threads check the memory map and see it empty (and then proceed to do the reorder)
the check memory map needs to be done while holding lock. 
please ensure you build with --enable_onnx_tests and tested on onnx models. 
seems like a bunch of models are failing in the CI run.

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -4,10 +4,12 @@
 #include "core/framework/allocatormgr.h"
 #include "core/framework/execution_provider.h"
 #include "core/graph/graph_transformer.h"
+#include "mkldnn.hpp"
```
forward declare mkldnn:memory instead of including mkldnn header. otherwise you leak mkldnn header to those that include mkldnn_execution_provider.h
did you test python bindings by building with --enable_pybind? the build there fails due to this issue.

sreekanth-yalachigere replied:
forward declared `mkldnn::memory`

```
#ifndef MKLDNN_HPP
namespace mkldnn {
class memory;
};
#endif
```

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +39,28 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
 
   virtual std::shared_ptr<KernelRegistry> GetKernelRegistry() const override;
+
+  std::shared_ptr<mkldnn::memory> Get_weight_memory(std::string weightName) {
```
please follow coding conventions:
https://github.com/Microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md

function names should be GetWeighstMemory(), SetWeightsMemory(), ReorderWeights() etc.

jywu-msft replied:
please also fix the style/formatting by using clang-format with google c++ style rules.

sreekanth-yalachigere replied:
enabled clang support in vs2017

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +39,28 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+  // mkldnn formatted weights(filer data) memory from first iteration 
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map;
+  mutable std::mutex mutex_;
```
you still haven't addressed issue of weight reordering happening only once (via mutex or std::call_once)

sreekanth-yalachigere replied:
please see the [comment](https://github.com/Microsoft/onnxruntime/pull/256#discussion_r244677367) above

jywu-msft commented on onnxruntime/core/providers/mkldnn/nn/conv.h:
```diff
@@ -4,18 +4,29 @@
 
   Status Compute(OpKernelContext* context) const override;
 
+std::shared_ptr<mkldnn::memory> Reorder_weights(OpKernelContext* context,
```
this method shouldn't be public.

pranavsharma commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,23 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+ private:
+  // mkldnn weights(filer data) memory blocks from first iteration
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map;
```
weights_mem_map naming convention is wrong.

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -402,20 +413,34 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
+    filter_dst_mem = provider_->GetWeightsMemory(weightKey);
+
+    if (filter_dst_mem == nullptr) {
+      std::lock_guard<std::mutex> lock(mutex_);
```
1. Is weights_mem_map going to be used only by this kernel?
2. Store the mutex along with the data structure that is guarding it.
3. The locking code below seems wrong since there can be races between checks and updates. Wrap the get/update code around a mutex to ensure thread safety.

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -254,6 +256,15 @@ class ConvPrimitivePool : public PrimitivePool<T> {
+  MemoryReorderParams params(src, dst);
+  DoReorder<T>(params);
+  std::string weight_name = OpKernel::Node().InputDefs()[1]->Name();
+  std::shared_ptr<mkldnn::memory> p = std::make_shared<mkldnn::memory>(dst);
```
why is this statement required?

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -402,20 +413,34 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
+    // Reorder filter memory layout if necessary
+    // Avoid data reordering. Save filter memory in mkldnn format from first iteration
+    // in execution provider mapped by weight name.
+    std::string weightKey = OpKernel::Node().InputDefs()[1]->Name();
```
const std::string& weight_key = .... or 
auto& weight_key = ...

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.h:
```diff
@@ -4,18 +4,31 @@
 namespace mkl_dnn {
+namespace {
+// Struct which encapsulates parameters for MKLDNN Conv primitive.
+struct ConvParams;
```
who needs this in this file?

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -254,6 +256,15 @@ class ConvPrimitivePool : public PrimitivePool<T> {
+void Conv<T>::ReoderFilterMemory(const mkldnn::memory& src, const mkldnn::memory& dst) const {
+  MemoryReorderParams params(src, dst);
+  DoReorder<T>(params);
+  std::string weight_name = OpKernel::Node().InputDefs()[1]->Name();
```
const std::string& weight_name = .... or
auto& weight_name = ...

pranavsharma commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,23 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+ private:
+  // mkldnn weights(filer data) memory blocks from first iteration
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map;
```
You don't need shared_ptr's. You can use unique_ptr and return raw ptrs.

sreekanth-yalachigere replied:
@pranavsharma @jywu-msft 

Tried the following and it works. But I had to

Remove forward_reference and include mkldnn.hpp. This will cause pybuild errors
```
namespace mkldnn {
class memory;
};
```

Convert raw-pointer to unique_ptr to Save memory in provider (conv.cc)
`std::unique_ptr<mkldnn::memory>{filter_dst_mem}`

const and reference doesn't work in SetWeightsMemory (mkldnn_provider.h)
    `weights_mem_map1_.emplace(weight_key, std::move(filter_dst_mem));`

Please suggest.

**mkldnn_provider.h**
```
  mkldnn::memory* GetWeightsMemory1(const std::string& weight_key) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto iter = weights_mem_map1_.find(weight_key);
    if (iter != weights_mem_map1_.end())
      return iter->second.get();
    return nullptr;
  }

  void SetWeightsMemory1(std::string weight_key,
                        std::unique_ptr<mkldnn::memory> filter_dst_mem) {
    std::lock_guard<std::mutex> lock(mutex_);
    weights_mem_map1_.emplace(weight_key, std::move(filter_dst_mem));
  }
  
 private:
  std::map<std::string, std::unique_ptr<mkldnn::memory>> weights_mem_map1_;

```
**conv.cc**
```
    mkldnn::memory* filter_dst_mem = nullptr;
    filter_dst_mem = provider_->GetWeightsMemory1(weight_name);

   if (filter_dst_mem == nullptr) {
      if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
        auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(
                                                     filter_dims_mkl, MklDnnType<T>(), filter_format),
                                                 cpu_engine);
        mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
        filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
        filter_dst_mem = 
            new mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get());
        MemoryReorderParams params(src, *filter_dst_mem);
        DoReorder<T>(params);
        provider_->SetWeightsMemory1(weight_name, std::unique_ptr<mkldnn::memory>{filter_dst_mem});
        filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
      }
    } else {
      filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
    }
```

pranavsharma replied:
Why did you have to remove the forward reference? Doesn't keeping the forward ref work while using unique_ptr?

sreekanth-yalachigere replied:
I can not compile otherwise. :-(

pranavsharma commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -4,11 +4,18 @@
 #include "core/framework/execution_provider.h"
 #include "core/graph/graph_transformer.h"
 
+#ifndef MKLDNN_HPP
```
do we need this?

sreekanth-yalachigere replied:
Yes. We are forward defining `mkldnn::memory` `py_byild` (`--enable_pybind`) was failing. 
When I include this in `conv.h`, it was giving compilation error as it also has `mkldnn.hpp` included.

pranavsharma replied:
i understand that you're forward defining 'memory'. i didn't understand the ifndef part. why is that needed?

sreekanth-yalachigere replied:
conv.cc gets mkldnn::memory definition from both mkldnn_provider.h and mkldnn.hpp and we get compilation error. ifndef MKLDNN_HPP makes conv.cc ignore forward definition from mkldnn_provider.h

pranavsharma replied:
will look into this more. sounds like a hack.

sreekanth-yalachigere replied:
Removed #ifndef.

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -402,20 +413,34 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
+    // in execution provider mapped by weight name.
+    std::string weightKey = OpKernel::Node().InputDefs()[1]->Name();
+    std::shared_ptr<mkldnn::memory> filter_dst_mem = nullptr;
+    filter_dst_mem = provider_->GetWeightsMemory(weightKey);
```
This should be protected by a lock.

sreekanth-yalachigere replied:
moved mutex to provider class.

pranavsharma commented:
@sreekanth-yalachigere  are you done with all the changes suggested in the review?

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,23 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+    return nullptr;
+  }
+
+  void SetWeightsMemory(std::string weight_key,
```
implemented.

pranavsharma replied:
i still see weight_key passed by value.

sreekanth-yalachigere replied:
:-( will fix it.

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,23 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
 
   virtual std::shared_ptr<KernelRegistry> GetKernelRegistry() const override;
+
+  std::shared_ptr<mkldnn::memory> GetWeightsMemory(std::string weight_key) {
```
implemented.

sreekanth-yalachigere commented:
Yes. Committed changes last night but wanted to  make sure all tests and CI backend test data pass.

Moved `mutex` to provider class and `lock_gaurd` at both `Get `and `Set`
passing memory and weight key as const ref

Re: Shared pointers for weights_mem_map_
Since we are forward defining mkldnn::memory, I get compilation errors with unique_ptrs. Since our intent is to share this memory with other threads, I am wondering if we can keep shared_ptrs.

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/conv.cc:
```diff
@@ -402,20 +404,29 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
-      MemoryReorderParams params(src, dst);
-      DoReorder<T>(params);
-      filter_data = static_cast<T*>(dst.get_data_handle());
+    // Reorder filter memory layout if necessary
```
The code below can cause the reordering to take place more than once. This doesn't address the requirement to do the reordering only once. You should acquire the lock before calling GetWeightsMemory and release it after SetWeightsMemory.

pranavsharma commented:
please resolve build failures.

sreekanth-yalachigere commented:
@pranavsharma I forked main repo on 21st. I just merged main repo to my branch. Can you please kick off CI pipeline builds again?

jywu-msft commented:
most of the model tests fail using mkldnn provider with your changes. I had raised this issue before. 
Can you please investigate?

2019-01-07T18:36:42.6168709Z test densenet121 failed, please fix it
2019-01-07T18:36:42.6169226Z test bvlc_reference_caffenet failed, please fix it
2019-01-07T18:36:42.6171265Z test zfnet512 failed, please fix it
2019-01-07T18:36:42.6173040Z test tiny_yolov2 failed, please fix it
2019-01-07T18:36:42.6173419Z test emotion_ferplus failed, please fix it
2019-01-07T18:36:42.6176021Z test bvlc_reference_rcnn_ilsvrc13 failed, please fix it
2019-01-07T18:36:42.6177070Z test vgg19 failed, please fix it
2019-01-07T18:36:42.6177223Z test mnist failed, please fix it
2019-01-07T18:36:42.6179460Z test inception_v2 failed, please fix it
2019-01-07T18:36:42.6180698Z test inception_v1 failed, please fix it
2019-01-07T18:36:42.6181971Z test shufflenet failed, please fix it
2019-01-07T18:36:42.6183297Z test resnet18v2 failed, please fix it
2019-01-07T18:36:42.6184892Z test mobilenetv2-1.0 failed, please fix it
2019-01-07T18:36:42.6185271Z test bvlc_alexnet failed, please fix it
2019-01-07T18:36:42.6187854Z test resnet34v2 failed, please fix it
2019-01-07T18:36:42.6203349Z test squeezenet1.1 failed, please fix it
2019-01-07T18:36:42.6203870Z test resnet50v2 failed, please fix it
2019-01-07T18:36:42.6204256Z test resnet50 failed, please fix it
2019-01-07T18:36:42.6204616Z test squeezenet failed, please fix it
2019-01-07T18:36:42.6204789Z test resnet101v2 failed, please fix it
2019-01-07T18:36:42.6207768Z test resnet152v2 failed, please fix it
2019-01-07T18:36:42.6208200Z test bvlc_googlenet failed, please fix it
2019-01-07T18:36:42.8522988Z result:

sreekanth-yalachigere commented:
Problem root caused and fixed. Memory allocated to reordered buffers was going out of scope and all reordered data was lost.  
To keep all allocations in scope, created list in provider class and added SaveAllocatedMemory member function.
```
  std::list<IAllocatorUniquePtr<void>> reordered_buffers_;
  void SaveAllocatedMemory(IAllocatorUniquePtr<void> buffer);

```
from conv.cc, called provider SaveAllocatedMemory after reordering to keep buffer in scope.

thanks

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,38 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
 
   virtual std::shared_ptr<KernelRegistry> GetKernelRegistry() const override;
+
+  std::shared_ptr<mkldnn::memory> GetWeightsMemory(const std::string& weight_key) {
```
can we make GetWeightsMemory(), SetWeightsMemory(), GetMutex(), and SaveAllocatedMemory() private
they shouldn't be exposed externally. 

and we can get rid of the lock_guard in GetWeightsMemory and SetWeightsMemory

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,38 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map_;
+  // Save reordered memory buffers in list so that memory is not freed.
+  std::list<IAllocatorUniquePtr<void>> reordered_buffers_;
```
This works.
ideally, it would have been nice to wrap the mkldnn::memory descriptor and buffer in another structure since they go together. but we can consider that refactor in the future.

jywu-msft replied:
btw, why list instead of vector?

sreekanth-yalachigere replied:
replaced list with vector.

jywu-msft commented on onnxruntime/core/providers/mkldnn/nn/conv.h:
```diff
@@ -4,18 +4,23 @@
 #pragma once
 #include "core/framework/op_kernel.h"
 #include "core/providers/cpu/nn/conv.h"
+#include "../mkldnn_execution_provider.h"
```
better to not include as a relative path.

sreekanth-yalachigere replied:
fixed.

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,38 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+ private:
+  // mkldnn weights(filer data) memory blocks from first iteration
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map_;
```
why map instead of unordered_map?

sreekanth-yalachigere replied:
yes. in this case we can use unordered_map.

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,38 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+    return conv_mutex_;
+  }
+
+  void SaveAllocatedMemory(IAllocatorUniquePtr<void> buffer) {
```
prefer a more descriptive name. maybe SaveWeightsMemoryBuffer

sreekanth-yalachigere replied:
Done. SaveWeightsMemoryBuffer and GetWeightsMemoryBuffer

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h:
```diff
@@ -37,6 +44,35 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
+  std::unordered_map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map_;
+  // Save reordered memory buffers in list so that memory is not freed.
+  std::vector<IAllocatorUniquePtr<void>> reordered_buffers_;
+  std::mutex conv_mutex_;
```
renamed it to `mutex_`
